cf-custom-resources/lib/bucket-cleaner.js

```
@@ -0,0 +1,191 @@
+    Bucket: bucketName
+  }
+  while (true) {
+    const listResp = await s3.listObjectVersions(listObjectVersionsParam).promise();
+    // After deleting other versions, remove delete markers version.
+    // For info on "delete marker": https://docs.aws.amazon.com/AmazonS3/latest/dev/DeleteMarker.html
+    let objectsToDelete = [
+      ...listResp.Versions.map(version => ({ Key: version.Key, VersionId: version.VersionId })),
+      ...listResp.DeleteMarkers.map(marker => ({ Key: marker.Key, VersionId: marker.VersionId }))
+    ];
+    if (objectsToDelete.length === 0) {
+      return
+    }
+    const delResp = await s3.deleteObjects({
+      Bucket: bucketName,
+      Delete: {
+        Objects: objectsToDelete,
+        Quiet: true
+      }
+    }).promise()
+    if (delResp.Errors.length > 0) {
+      throw new AggregateError([new Error(`${delResp.Errors.length}/${objectsToDelete.length} objects failed to delete`),
+      new Error(`first failed on key "${delResp.Errors[0].Key}": ${delResp.Errors[0].Message}`)]);
+    }
+    if (!listResp.IsTruncated) {
+      return
+    }
+    listObjectVersionsParam.KeyMarker = listResp.NextKeyMarker
+    listObjectVersionsParam.VersionIdMarker = listResp.NextVersionIdMarker
+  }
+};
+
+/**
+ * Correct desired count handler, invoked by Lambda.
+ */
+exports.handler = async function (event, context) {
+  var responseData = {};
+  const props = event.ResourceProperties;
+  const physicalResourceId = event.PhysicalResourceId || `bucket-cleaner-${event.LogicalResourceId}`;
+
+  try {
+    switch (event.RequestType) {
+      case "Create":
+      case "Update":
+        break;
+      case "Delete":
+        await cleanBucket(props.BucketName);
+        break;
+      default:
+        throw new Error(`Unsupported request type ${event.RequestType}`);
+    }
+    await report(event, context, "SUCCESS", physicalResourceId, responseData);
+  } catch (err) {
+    console.log(`Caught error ${err}.`);
```

should this be a best-effort clean up? i.e. if there is an error, we log it, but don't fail the operation.

---

We might want to surface the error to users if the custom resource fails, because it's less likely for this custom resource to fail for random reasons 💭 This is what we are doing to our stackset bucket and static site bucket too.